console/collections/src/merkle_tree/mod.rs

```diff
@@ -213,6 +213,20 @@ impl<E: Environment, LH: LeafHash<Hash = PH::Hash>, PH: PathHash<Hash = Field<E>
     pub fn update(&mut self, leaf_index: usize, new_leaf: &LH::Leaf) -> Result<()> {
         let timer = timer!("MerkleTree::update");
 
+        // Compute the updated Merkle tree with the new leaves.
+        let updated_tree = self.prepare_update(leaf_index, new_leaf)?;
+        // Update the tree at the very end, so the original tree is not altered in case of failure.
+        *self = updated_tree;
```

Comment:
@ljedrz can you check efficiency and safety here? We want this to first build a candidate tree, fail if it needs to, then update in place if it is successful.

Reply:
It's fine in terms of avoiding mutation in case anything goes wrong and only replacing the original object in case of success.
As for efficiency, technically the cloning of the original tree doesn't need to happen - note that the operations starting from the `// Initialize the Merkle tree.` comment are all infallible, so at that point the original tree might as well be updated in place.